cupy/lib/polynomial.pyx

```diff
@@ -0,0 +1,275 @@
+    line2 = ''
+    output = ' '
+    while True:
+        mat = _poly_mat.search(astr, n)
+        if mat is None:
+            break
+        span = mat.span()
+        power = mat.groups()[0]
+        partstr = astr[n:span[0]]
+        n = span[1]
+        toadd2 = partstr + ' ' * (len(power) - 1)
+        toadd1 = ' ' * (len(partstr) - 1) + power
+        if ((len(line2) + len(toadd2) > wrap) or
+                (len(line1) + len(toadd1) > wrap)):
+            output += line1 + '\n' + line2 + '\n '
+            line1 = toadd1
+            line2 = toadd2
+        else:
+            line2 += partstr + ' ' * (len(power) - 1)
+            line1 += ' ' * (len(partstr) - 1) + power
+    output += line1 + '\n' + line2
+    return output + astr[n:]
+
+
+class poly1d(object):
+    """A one-dimensional polynomial class.
+
+    Args:
+        c_or_r (cupy.ndarray or cupy.poly1d): The polynomial's
+         coefficients in decreasing powers
+        r (bool, optional): If True, `c_or_r` specifies the
+            polynomial's roots; the default is False.
+        variable (str, optional): Changes the variable used when
+            printing the polynomial from `x` to `variable`
+
+    .. seealso:: :func:`numpy.poly1d`
+
+    """
+    __hash__ = None
+
+    @property
+    def coeffs(self):
+        return self._coeffs
+
+    @coeffs.setter
+    def coeffs(self, value):
+        if value is not self._coeffs:
+            raise AttributeError('Cannot set attribute')
+
+    @property
+    def variable(self):
+        return self._variable
+
+    @property
+    def order(self):
+        return self._coeffs.size - 1
+
+    # TODO
+    @property
+    def roots(self):
+        pass
+
+    @property
+    def _coeffs(self):
+        return self.__dict__['coeffs']
+
+    @_coeffs.setter
+    def _coeffs(self, coeffs):
+        self.__dict__['coeffs'] = coeffs
+
+    r = roots
+    c = coef = coefficients = coeffs
+    o = order
+
+    def __init__(self, c_or_r, r=False, variable=None):
+        if isinstance(c_or_r, poly1d):
+            self._variable = c_or_r._variable
+            self._coeffs = c_or_r._coeffs
+
+            if set(c_or_r.__dict__) - set(self.__dict__):
+                msg = ('In the future extra properties will not be copied '
+                       'across when constructing one poly1d from another')
+                warnings.warn(msg, FutureWarning, stacklevel=2)
+                self.__dict__.update(c_or_r.__dict__)
+
+            if variable is not None:
+                self._variable = variable
+            return
+        # TODO: if r: c_or_r = poly(c_or_r)
+        if c_or_r.ndim > 1:
+            raise ValueError('Polynomial must be 1d only.')
+        c_or_r = cupy.trim_zeros(c_or_r, trim='f')
+        if c_or_r.size == 0:
+            c_or_r = cupy.array([0.])
+        self._coeffs = c_or_r
+        if variable is None:
+            variable = 'x'
+        self._variable = variable
+
+    def __array__(self, t=None):
```

Comment:
We don't want to support `__array__` method to avoid implicit type conversion. Please fix to raise `TypeError` like `cupy.core.core.ndarray.__array__`.